sdb/commands/stacks.py

```diff
@@ -216,120 +272,106 @@ def find_module_memory_segment(mod_name: str) -> Tuple[int, int]:
+                #
+                func = sdb.get_object(self.args.function)
+                sym = sdb.get_symbol(func.address_of_())
+            except KeyError:
+                raise sdb.CommandError(
+                    self.name,
+                    "symbol '{:s}' does not exist".format(self.args.function))
+            if func.type_.kind != drgn.TypeKind.FUNCTION:
+                raise sdb.CommandError(
+                    self.name,
+                    "'{:s}' is not a function".format(self.args.function))
+            self.func_start = sym.address
+            self.func_end = self.func_start + sym.size
 
-        #
-        # Resolve TSTATE shortcut and/or sanitize it to standard uppercase
-        # notation if it exists.
-        #
         if self.args.tstate:
-            if self.args.tstate in Stacks.TASK_STATE_SHORTCUTS:
-                self.args.tstate = Stacks.TASK_STATES[
-                    Stacks.TASK_STATE_SHORTCUTS[self.args.tstate]]
-            else:
-                self.args.tstate = self.args.tstate.upper()
+            self.match_state = Stacks.resolve_state(self.args.tstate)
+            task_states = Stacks.TASK_STATES.values()
+            if self.match_state not in task_states:
+                raise sdb.CommandError(
+                    self.name,
+                    "'{:s}' is not a valid task state (acceptable states: {:s})"
+                    .format(self.args.tstate, ", ".join(task_states)))
 
-        mod_start, mod_end = -1, -1
         if self.args.module:
-            mod_start, mod_size = Stacks.find_module_memory_segment(
+            if Stacks.find_module_memory_segment(self.args.module)[0] == -1:
+                raise sdb.CommandError(
+                    self.name,
+                    "module '{:s}' doesn't exist or isn't currently loaded".
+                    format(self.args.module))
+            self.mod_start, mod_size = Stacks.find_module_memory_segment(
                 self.args.module)
-            assert mod_start != -1
-            mod_end = mod_start + mod_size
-
+            assert self.mod_start != -1
+            self.mod_end = self.mod_start + mod_size
+
+    def match_stack(self, task: drgn.Object) -> bool:
+        task_state = Stacks.task_struct_get_state(task)
+        if self.args.tstate and self.match_state != task_state:
+            return False
+
+        frame_pcs = Stacks.get_frame_pcs(task)
+        mod_match = False
+        func_match = False
+        for frame_pc in frame_pcs:
+            if self.args.module and self.mod_start <= frame_pc < self.mod_end:
+                mod_match = True
+
+            if self.args.function and self.func_start <= frame_pc < self.func_end:
+                func_match = True
+
+        if (self.args.function and not func_match) or (self.args.module and
+                                                       not mod_match):
+            return False
+        return True
```

Comment:
Not sure if that'd be too much but we could short-circuit the flow here a bit. I thought of 2 things:

[1] Iterate through the frames only if we're actually asked to filter by module or function.
[2] If everything has matched up to now, don't bother looking at subsequent frames.

```
    def match_stack(self, task: drgn.Object) -> bool:
        if self.args.tstate and self.match_state != Stacks.task_struct_get_state(task):
            return False

        if not (self.args.module or self.args.function): # point [1]
            return True

        mod_match, func_match = not self.args.module, not self.args.function # point [2]
        for frame_pc in Stacks.get_frame_pcs(task):
            if not mod_match and self.mod_start <= frame_pc < self.mod_end:
                mod_match = True

            if not func_match and self.func_start <= frame_pc < self.func_end:
                func_match = True

            if mod_match and func_match:
                return True
        return False
```

Probably splitting hairs here so feel free to drop any/both of the above.